server.py

```diff
@@ -36,17 +36,25 @@ def index():
 ######################################################################
 @app.route('/products', methods=['GET'])
 def list_products():
-    category        = request.args.get('category')
-    discontinued    = str2bool(request.args.get('discontinued'))
-    results     = []
-    for key, value in products.iteritems():
-        if matchesClause(category,      value['category']) and \
-           matchesClause(discontinued,  value['discontinued']):
-             results.append(products[key])
-    return reply(results, HTTP_200_OK)
+	category=request.args.get('category')
+	discontinued = str2bool(request.args.get('discontinued'))
+	search_term=(request.args.get('search')).lower()
+	#print ("Category", category)
+	#print ("discontinued", discontinued)
+	#print ("search_term", search_term)
+	results = []
+	for key, value in products.iteritems():
+		if category is None and discontinued is None and search_term is None:
+			results.append(products[key])
+		elif matchesClause(category, value['category']) \
+			and matchesClause(discontinued, value['discontinued']):
+			results.append(products[key])
+		elif search_term in value['category'].lower() or search_term in value['name'].lower():
+			results.append(products[key])
+	return reply(results, HTTP_200_OK)
 
 def matchesClause(clause_value, item_value):
-    return clause_value is None or clause_value == item_value 
+    return clause_value is not None and clause_value in item_value 
```

Comment:
The `matchesClause` method should not have been modified.

1. `is not None`: What if no URL parameter is passed by the user? The method will return false, which is undesired. I believe it is for this reason that the long condition in line 47 was added, but again this has issues because: a) it's long , hard to read, and makes for highly-coupled logic, whereas the helper-method approach described above is simpler, b) what if the user passes a single URL attribute but not all of them? I don't know what the behavior would be - but I doubt it would function as desired
2. `'=='` was replaced with `'in'`, which at face-value would seem to add to the flexibility of the code, as the second parameter can be an Iterable type. But it breaks the code, e.g. checking for a `bool` value, such as 'discontinued' attribute, will throw the following error:
`TypeError: argument of type 'bool' is not iterable`. Furthermore all of the product attributes (name, description, etc.) are scalar values, not lists (the string type kind of asks as both but even so) so there is no mandate to do this.

If the check for a URL parameter value requires more involved logic for which the `matchesClause` method does not suffice, such as the one you are implementing for the search term, then add another helper method to handle such a case.